services/kafka/hash.go

```diff
@@ -0,0 +1,116 @@
+package kafka
+
+import (
+	"errors"
+	"hash/crc32"
+	"math/rand"
+	"time"
+
+	"github.com/Shopify/sarama"
+)
+
+var ErrNonPositivePartitions = errors.New("number of partitions must be positive")
+
+type murmur2 struct{}
+
+// Partition here is based on the murmur2 balance function in github.com/segmentio/kafka-go to not break compatibility
+// with earlier versions.
+func (b murmur2) Partition(msg *sarama.ProducerMessage, numPartitions int32) (int32, error) {
+	if numPartitions == 0 {
+		return 0, ErrNonPositivePartitions
+	}
+
+	// NOTE: the murmur2 balancers in java and librdkafka treat a nil key as
+	//       non-existent while treating an empty slice as a defined value.
+	if msg.Key == nil {
+		return rand.New(rand.NewSource(time.Now().UTC().UnixNano())).Int31n(numPartitions), nil
+	}
+	key, err := msg.Key.Encode()
+	if err != nil {
+		return 0, err
+	}
+	return int32((murmur2Hash(key) & 0x7fffffff) % uint32(numPartitions)), nil
+}
+
+func (murmur2) RequiresConsistency() bool { return true }
+
+// murmur2 is taken from github.com/segmentio/kafka-go
+// We copy here because it was in a private method.
+// It is a go port of the Java library's murmur2 function.
+// https://github.com/apache/kafka/blob/1.0/clients/src/main/java/org/apache/kafka/common/utils/Utils.java#L353
+// Go port of the Java library's murmur2 function.
+// https://github.com/apache/kafka/blob/1.0/clients/src/main/java/org/apache/kafka/common/utils/Utils.java#L353
+func murmur2Hash(data []byte) uint32 {
```

Comment:
Does the Java library have any murmur2 test vectors we can use to make sure our port is correct?